specifyweb/frontend/js_src/lib/components/Notifications/hooks.tsx

```diff
@@ -0,0 +1,161 @@
+import type { LocalizedString } from 'typesafe-i18n';
+
+import { ajax } from '../../utils/ajax';
+import { formatDateForBackEnd } from '../../utils/parser/dateFormat';
+import type { IR, RA } from '../../utils/types';
+import { sortFunction } from '../../utils/utils';
+import { formatUrl } from '../Router/queryString';
+import type { GenericNotification } from './NotificationRenderers';
+
+const INITIAL_INTERVAL = 5000;
+const INTERVAL_MULTIPLIER = 1.1;
+let addSinceToNotification = true;
+
+export function useNotificationsFetch({
+  freezeFetchPromise,
+  isOpen,
+}: {
+  readonly freezeFetchPromise: React.MutableRefObject<
+    Promise<void> | undefined
+  >;
+  readonly isOpen: boolean;
+}): {
+  readonly notifications: RA<GenericNotification> | undefined;
+  readonly setNotifications: React.Dispatch<
+    React.SetStateAction<RA<GenericNotification> | undefined>
+  >;
+} {
+  const [notifications, setNotifications] = React.useState<
+    RA<GenericNotification> | undefined
+  >(undefined);
+  React.useEffect(() => {
+    let pullInterval = INITIAL_INTERVAL;
+    let lastFetchedTimestamp: Date | undefined;
+    let timeout: NodeJS.Timeout | undefined = undefined;
+
+    const doFetch = (since = new Date()): void => {
+      const startFetchTimestamp = new Date();
+
+      const url = addSinceToNotification
+        ? getSinceUrl(`/notifications/messages/`, since)
+        : `/notifications/messages/`;
+
+      /*
+       * Poll interval is scaled exponentially to reduce requests if the tab is
+       * left open.
+       */
+      pullInterval *= INTERVAL_MULTIPLIER;
+
+      // Don't fetch while a message is being deleted or marked as read
+      (freezeFetchPromise.current ?? Promise.resolve())
+        .then(async () =>
+          ajax<
+            RA<
+              Omit<GenericNotification, 'messageId' | 'payload'> & {
+                // eslint-disable-next-line @typescript-eslint/naming-convention
+                readonly message_id: string;
+              }
+            >
+          >(url, {
+            headers: { Accept: 'application/json' },
+            /*
+             * Don't show modal error dialog on errors. Several reasons why:
+             *  - Request may fail if Django migrations weren't run
+             *  - Request is initialized automatically, not by user, thus having
+             *    an error dialog appear out of blue could be confusing/unexpected
+             *  - Notifications is not a critical component, so if it fails, it
+             *    shouldn't bring down entire application
+             */
+            errorMode: 'silent',
+          })
+        )
+        .then(({ data: newNotifications }) => {
+          addSinceToNotification = false;
+
+          if (destructorCalled) return;
+
+          setNotifications((existingNotifications) =>
+            mergeAndSortNotifications(existingNotifications, newNotifications)
+          );
+
+          lastFetchedTimestamp = startFetchTimestamp;
```

Comment:
this should probably also be stored in a ref - right now lastFetchedTimestamp value is lost on useEffect re-executions (happens every time the user opens/closes the notifications component)

Reply:
this one is simpler to explain.

lastFetchedTimestamp stores the timestamp of last fetch
yet this is a variable you created in useEffect
so every time useEffect re-runs, the value of this variable is lost, and new variable is created
and useEffect has `[isOpen]` dependency array, thus it re-runs every time user opens or closes notifications panel

but it is not right to clear the value of that variable on user open/close of notifications as that shouldn't be related to lastFetchedTimestamp. Instead, lastFetchedTimestamp is related to "notifications" state (since lastFetchedTimestamp roughly describes when notifications state was last updated), so this variable should have the same lifecycle as the "notifications" hook